### PR TITLE
Certificate  CRUD 

### PR DESCRIPTION
**Pull request by jackmangosa1**
### #### PR description
This PR implements backend endpoints for creating, updating, and deleting certificates. 

## Checklist:
- [x] Created backend endpoint for adding new certificates.
- [x] Implemented backend endpoint to update existing certificates.
- [x]  Added backend endpoint to delete certificates.
- [x]  Included validation for input data in certificate operations.
- [x]  Added proper error handling for edge cases in create, update, and delete operations.
- [x] Updated the documentation to reflect the new certificate management endpoints.
 

#### How To Test Changes

- Clone the repository: Clone the repository:
  ```bash
  git clone https://github.com/jackmangosa1/certificate-manager.git

 #### Running the frontend
- Navigate to the frontend folder::   
```bash
   cd frontend
```  
- Install dependencies:   
```bash
   npm install
```
- Run the application locally:
```bash
npm start
```
### Running the backend
- Navigate to the backend folder::   
```bash
   cd backend
```  
- Install dependencies:   
```bash
   dotnet restore
```
- Run the application locally:   
```bash
   dotnet run
```

### CODE REVIEW

File: backend/Controllers/CertificateController.cs [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+using Microsoft.AspNetCore.Mvc;
+
+namespace CertificateMangerAPI.Controllers
+{
+    [Route("api/certificates")]
+    [ApiController]
+    public class CertificateController : ControllerBase
+    {
+        private readonly ICertificateService _certificateService;
+        private readonly IMapper _mapper;
```
Comment (elhasic) [inline backend/Controllers/CertificateController.cs:14]: Put here new line

File: backend/Controllers/CertificateController.cs [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        }
+
+        [HttpGet]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        public async Task<ActionResult<IEnumerable<CertificateSummaryDTO>>> GetAllCertificates()
+        {
+            var certificates = await _certificateService.GetAllCertificatesAsync();
+            return Ok(certificates);
+        }
+
```
Comment (elhasic) [inline backend/Controllers/CertificateController.cs:49]: Remove unnecessary line

File: backend/Controllers/CertificateController.cs [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+            }
+
+            var certificate = await _certificateService.GetCertificateByIdAsync(id);
+            if (certificate == null)
+            {
+                return NotFound("Certificate is not found");
+            }
+
+            var certificateToPatch = _mapper.Map<UpdateCertficateDTO>(certificate);
+
```
Comment (elhasic) [inline backend/Controllers/CertificateController.cs]: Remove unnecessary line

File: backend/Data/CertificateManagerDbContext.cs [hunk 4]
```diff
@@ -30,7 +30,8 @@ public CertificateManagerDbContext(DbContextOptions<CertificateManagerDbContext>
 
     public virtual DbSet<User> Users { get; set; }
 
-
+    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
```
Comment (elhasic) [inline backend/Data/CertificateManagerDbContext.cs]: This method should not use here, all connection string configuration should be in Program.cs

File: backend/Mappers/MappingProfile.cs [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+namespace CertificateMangerAPI.Mappers
+{
+    public class MappingProfile : Profile
+    {
+        public MappingProfile()
+        {
+            CreateMap<string, DateOnly>().ConvertUsing<StringToDateOnlyConverter>();
+
+            CreateMap<Certificate, CreateCertificateDTO>()
```
Comment (elhasic) [inline backend/Mappers/MappingProfile.cs]: Prepare separate Mapper for each DTO

File: backend/Repositories/CertificateRepository.cs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+            await _context.SaveChangesAsync();
+            return _mapper.Map<CreateCertificateDTO>(certificate);
+        }
+
+        public async Task<CreateCertificateDTO> GetCertificateByIdAsync(int certificateId)
+        {
+            var certificate = await _context.Certificates.FindAsync(certificateId);
+            return _mapper.Map<CreateCertificateDTO>(certificate);
+        }
+
```
Comment (elhasic) [inline backend/Repositories/CertificateRepository.cs:31]: Remove unnecessary line

File: backend/Repositories/CertificateRepository.cs [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        public async Task<IEnumerable<CertificateSummaryDTO>> GetAllCertificatesAsync()
+        {
+            var certificates = await _context.Certificates
+                .Include(c => c.Supplier)
+                .Include(c => c.CertificateType)
+                .ToListAsync();
+
+            return _mapper.Map<IEnumerable<CertificateSummaryDTO>>(certificates);
+        }
+
```
Comment (elhasic) [inline backend/Repositories/CertificateRepository.cs:41]: Remove unnecessary lines

File: backend/Repositories/CertificateRepository.cs [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+            if (certificate == null)
+            {
+                throw new KeyNotFoundException($"Certificate with ID {certificateId} not found.");
+            }
+
+            _mapper.Map(certificateDTO, certificate);
+
+            await _context.SaveChangesAsync();
+        }
+
```
Comment (elhasic) [inline backend/Repositories/CertificateRepository.cs:54]: Remove unnecessary line

File: backend/Services/CertificateService.cs [hunk 9]
```diff
@@ -0,0 +1,42 @@
+﻿using CertificateManagerAPI.DTO;
+using CertificateManagerAPI.Repositories;
+
+namespace CertificateManagerAPI.Services
+{
+    public class CertificateService : ICertificateService
+    {
+        private readonly ICertificateRepository _certificateRepository;
+        public CertificateService(ICertificateRepository certificateRepository)
```
Comment (elhasic) [inline backend/Services/CertificateService.cs:9]: Add new line

File: backend/Services/CertificateService.cs [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+        {
+            _certificateRepository = certificateRepository;
+        }
+        public async Task<CreateCertificateDTO> CreateCertificateAsync(CreateCertificateDTO certificateDTO)
```
Comment (elhasic) [inline backend/Services/CertificateService.cs:14]: Add new line

File: backend/Services/CertificateService.cs [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+        {
+            var certificates = await _certificateRepository.GetAllCertificatesAsync();
+            return certificates;
+        }
+
+        public async Task UpdateCertificateAsync(int certificateId, UpdateCertficateDTO certificateDTO)
+        {
+            await _certificateRepository.UpdateCertificateAsync(certificateId, certificateDTO);
+        }
+
```
Comment (elhasic) [inline backend/Services/CertificateService.cs:35]: Remove unnecessary line

File: backend/Services/ICertificateService.cs [hunk 12]
```diff
@@ -0,0 +1,14 @@
+﻿using CertificateManagerAPI.DTO;
+
+namespace CertificateManagerAPI.Services
+{
+    public interface ICertificateService
+    {
+
```
Comment (elhasic) [inline backend/Services/ICertificateService.cs]: Remove unnecessary line

File: backend/Utilities/GlobalExceptionHandler.cs [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+            {
+                errorMessage += $" Exception details: {exception}";
+            }
+
+            httpContext.Response.StatusCode = statusCode;
+            var errorResponse = new
+            {
+                Status = statusCode,
+                Message = errorMessage
+            };
```
Comment (elhasic) [inline backend/Utilities/GlobalExceptionHandler.cs:61]: Add new line

File: backend/Utilities/GlobalExceptionHandler.cs [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+                statusCode = StatusCodes.Status400BadRequest;
+                errorMessage = "Invalid operation.";
+            }
+
+            if (httpContext.RequestServices.GetRequiredService<IHostEnvironment>().IsDevelopment())
+            {
+                errorMessage += $" Exception details: {exception}";
+            }
+
+            httpContext.Response.StatusCode = statusCode;
```
Comment (elhasic) [inline backend/Utilities/GlobalExceptionHandler.cs:55]: Add new line

File: backend/Mappers/DateOnlyJsonConverter.cs [hunk 15]
```diff
@@ -0,0 +1,21 @@
+﻿using System.Globalization;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace CertificateMangerAPI.Mappers
```
Comment (elhasic) [inline backend/Mappers/DateOnlyJsonConverter.cs]: Move this file to Utilities folder

File: backend/Mappers/StringToDateOnlyConverter.cs [hunk 16]
```diff
@@ -0,0 +1,16 @@
+﻿using AutoMapper;
+
+namespace CertificateMangerAPI.Mappers
```
Comment (elhasic) [inline backend/Mappers/StringToDateOnlyConverter.cs]: Move this file to Utilities folder

**Review by elhasic — CHANGES_REQUESTED** [review]

File: backend/Controllers/CertificateController.cs [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+            await _certificateService.UpdateCertificateAsync(id, certificateDTO);
+            return NoContent();
+        }
+
+        [HttpPatch("{id:int}")]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        [ProducesResponseType(StatusCodes.Status204NoContent)]
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
+        public async Task<IActionResult> PatchCertificate(int id, [FromBody] JsonPatchDocument<UpdateCertficateDTO> certificateUpdates)
```
Comment (elhasic) [inline backend/Controllers/CertificateController.cs]: Do you need JsonPatchDocument in this case

**Review by elhasic — APPROVED** [review]
